crates/wasi-common/wig/src/wasi.rs

```diff
@@ -193,13 +194,13 @@ pub fn define_struct(args: TokenStream) -> TokenStream {
                             #(#format_args),*
                         );
                         unsafe {
-                            let memory = match mem.get() {
-                                Ok(e) => e,
+                            let mut memory = match mem.get() {
+                                Ok(e) => wiggle_runtime::GuestMemory::new(e.as_mut_ptr(), e.len() as u32),
```

Comment:
If we're designing the wiggle API to last (which I suspect we are!) then I think we need to be really careful about how memory is handled. For example I don't think this'll actually cut it (the current implementation) in the sense that I think it'll be best if we can hold on to the memory object/export rather than loading the pointer/length here. That way we don't have to worry about relocation during a call (e.g. `memory.grow`).

Another aspect is that we need to be extremely careful when touching guest memory, but I suspect the `GuestMemory` type is a good place we can audit all the unsafety and such with respect to this.

Reply:
BTW, feel free to submit any fixes, thoughts, issues, comments, etc., in [`wiggle`] as well. Ultimately, we'll transfer it into `wasmtime`, but in the meantime, feel free to browse there!

[`wiggle`]: https://github.com/kubkon/wiggle

Reply:
@alexcrichton those are good points. I'd love to have a detailed conversation about this stuff and iterate on the API design with you, but probably the wiggle crate is the right venue for now?You know where to find us.

Reply:
@alexcrichton I've now given you full write access to [`wiggle`](https://github.com/kubkon/wiggle) so feel free to tweak the code as you see fit.

Reply:
Ah yeah FWIW I moved most of this thought over to https://github.com/kubkon/wiggle/issues/32!